File: Entrega_Final/AlgoritmosScripts/t06_PolePlacement.py
import control.matlab as cntm
import matplotlib.pyplot as plt
import numpy as np

###########################################################################
###########################################################################
##### NO FUNCIONA - MATRICES A B C D DIFERENTES ENTRE MATLAB Y PYTHON #####
###########################################################################
###########################################################################
# Descripcion de la Planta
# Componentes planta RCRC
R1 = 10*1e3
C1 = 1*1e-6
R2 = 39*1e3
C2 = 1*1e-6

# Frecuencia / periodo de muestreo
fs = 200
h = 1/fs

# % Función de transferencia continua
nums = [1]
dens = [(R1*C1*R2*C2), (R1*C1+R1*C2+R2*C2), 1]
Hs = cntm.tf(nums, dens)
print(Hs)

# % Función de transferencia discreta
Hz = cntm.c2d(Hs, h, 'zoh')
[numz, denz] = cntm.tfdata(Hz)
print(Hz)

# % A, B, C, D Matrices espacio estado
ssS_sys = cntm.ss(Hs)
ssZ_sys = cntm.c2d(ssS_sys, h)


# % Polos del sistema en espacio continuo
pS_sys = cntm.pole(Hs)
print(f"Los polos del sistema son  {pS_sys[0]}, {pS_sys[1]}")


plt.figure(1)
cntm.pzmap(Hs, grid = True, title = 'Polos y ceros de H(s)')  # % Grafica los polos y ceros

plt.figure(2)
cntm.pzmap(Hz, grid = True, title = 'Polos y ceros de H(z)') # % Grafica los polos y ceros


###########################################################################
#  REQUERIMIENTOS

# % Aplicando requerimiento tr menor en 30%
poles_t = np.exp(pS_sys*1.3*h)

# % Obteniendo K
K = cntm.place(ssZ_sys.A, ssZ_sys.B, [poles_t[0],poles_t[1]])
print(f"Valor de K {K}")

# % Obteniendo K0
ssZcl_sys = cntm.ss(ssZ_sys.A - ssZ_sys.B*K, ssZ_sys.B, ssZ_sys.C, ssZ_sys.D)
gain_pp =  np.dot(np.dot(ssZcl_sys.C, np.linalg.inv((np.eye(2) - ssZcl_sys.A))), ssZcl_sys.B) 
K0 = 1/gain_pp
print(f"Valor de K0 {K0}")

# **********************************************
# Definir parámetros
t = np.arange(0, 2 + h, h)  # Intervalo de tiempo de 0 a 2 segundos con paso h

# Inicializar matrices para guardar resultados
x = np.zeros((ssZ_sys.A.shape[0], len(t)))
y = np.zeros(len(t))
u = np.ones(len(t))

print("Matrices Espacio estado necesarios")
print(ssZ_sys.A)
print(ssZ_sys.B)
print(ssZ_sys.C)
print(ssZ_sys.D)
 
# Bucle de simulación en espacio de estado
for i in range(len(t) - 1):
    x[:, i + 1] = (np.dot(ssZ_sys.A, x[:, i].reshape(-1, 1))  + np.dot(ssZ_sys.B, u[i]))[:,0]
    y[i] = np.dot(ssZ_sys.C, x[:, i].reshape(-1, 1))[0][0]

# Graficar resultados
plt.figure(3)
plt.plot(t[:-1], y[:-1])  # Se usa t[:-1] y y[:-1] para ajustar la longitud de t e y
plt.xlabel('Tiempo')
plt.ylabel('Salida y')
plt.title('Respuesta del sistema en espacio de estado')
plt.grid(True)


plt.show()

File: Entrega_Final/AlgoritmosScripts/t06_PolePlacement.m
%% Control Digital en Sistemas Embebidos - MSE - PRACTICA 4

% pkg load signal
% pkg load control

clc;
clear;
close all;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% Descripcion de la Planta
% Componentes planta RCRC
R1 = 10*1e3;
C1 = 1*1e-6;
R2 = 39*1e3;
C2 = 1*1e-6;

% Frecuencia / periodo de muestreo
fs = 200;
h = 1/fs;

% Función de transferencia continua
nums = [ 1 ];
dens = [(R1*C1*R2*C2) (R1*C1+R1*C2+R2*C2) 1];

% Valores obtenidos experimentalmente
nums = [ 1 ];
dens = [0.000216 0.08105 1.0004];
Hs = tf(nums, dens)

% Función de transferencia discreta
Hz = c2d(Hs, h, 'zoh')
[numz, denz] = tfdata(Hz, 'v');

% A, B, C, D Matrices espacio estado
ssS_sys = ss(Hs);
ssZ_sys = c2d(ssS_sys, h);
% [ssS_Phi, ssS_Gamma, ssS_C, ssS_D] = tf2ss(nums, dens); % <<<< esta mal 

% Polos del sistema en espacio continuo
pS_sys = pole(Hs);
msg = sprintf('Los polos del sistema son %f, %f',pS_sys(1),pS_sys(2));
disp(msg);

figure(1);
pzmap(Hs)   % Grafica los polos (y ceros si los hubiera)
grid on 
title('Diagrama de Polos del Sistema');

%% REQUERIMIENTOS
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% Aplicando requerimiento tr menor en 30%
poles_t = exp(pS_sys*1.3*h);

% Obteniendo K
K = place(ssZ_sys.A, ssZ_sys.B, [poles_t(1),poles_t(2)]);

% Obteniendo K0
ssZcl_sys = ss(ssZ_sys.A - ssZ_sys.B*K, ssZ_sys.B, ssZ_sys.C, ssZ_sys.D);
gain_pp = ssZcl_sys.C * (eye(2) - ssZcl_sys.A) ^ (-1) * ssZcl_sys.B;
K0 = 1/gain_pp;

% Obteniendo Integral

Ae  = [ssZ_sys.A zeros(length(ssZ_sys.A),1);-ssZ_sys.C 1];
Beu = [ssZ_sys.B;0];
Ber = [zeros(length(ssZ_sys.B),1);1];
Ce  = [ssZ_sys.C 0];

pdi = exp(-10*max(abs(pS_sys))*h);  

K1 = place(Ae,Beu,[poles_t;pdi]);

K1_a = K1(1:2)
K1_b = K1(3)

%%
% APLICACION - SIMULACION LAZO CERRRADO CON GANANCIA

t = 0:h:5;
fsig = 1;

x = zeros(length(ssZ_sys.A),length(t));
y = zeros(1,length(t));
u = ones(1,length(t));

for i=1:length(t)
    x(:,i+1) = ssZ_sys.A*x(:,i) + ssZ_sys.B*u(i);
    y(i) = ssZ_sys.C*x(:,i);
end

%
r    = ones(1,length(t));

r = square(2 * pi * fsig * t);
r = (r + 1) / 2 + 1;


x_lcg = zeros(length(ssZ_sys.A),length(t));
y_lcg = zeros(1,length(t));

for i=1:length(t)
    x_lcg(:,i+1) = ssZ_sys.A * x_lcg(:,i) - ssZ_sys.B * K * x_lcg(:,i) + ssZ_sys.B * K0 * r(i);
    y_lcg(i) = ssZ_sys.C * x_lcg(:,i);
end

figure(2)
plot(t,r,t,y_lcg,'LineWidth',1)